daxa/process/erosita/assemble.py

```diff
@@ -0,0 +1,202 @@
+    if (pattern <= 0 or pattern >= 16):
+        raise ValueError("Valid eROSITA patterns are between 1 and 15 inclusive")
+    
+    # Converting the parameters to the correct format for the esass command
+    lo_en = lo_en.value
+    hi_en = hi_en.value
+    if flag_invert:
+        flag_invert = 'yes'
+    else:
+        flag_invert = 'no'
+
+    # Sets up storage dictionaries for bash commands, final file paths (to check they exist at the end), and any
+    #  extra information that might be useful to provide to the next step in the generation process
+    miss_cmds = {}
+    miss_final_paths = {}
+    miss_extras = {}
+
+    # Just grabs the eROSITA missions, we already know there will be at least one because otherwise _esass_process_setup
+    #  would have thrown an error
+    erosita_miss = [mission for mission in obs_archive if mission.name in ALLOWED_EROSITA_MISSIONS]
+    # We are iterating through erosita missions (options could include erosita_cal_pv for instance).
+    for miss in erosita_miss:
+        # Sets up the top level keys (mission name) in our storage dictionaries
+        miss_cmds[miss.name] = {}
+        miss_final_paths[miss.name] = {}
+        miss_extras[miss.name] = {}
+
+        # This method will fetch the valid data (ObsID, Instruments) that can be processed
+        all_obs_info = obs_archive.get_obs_to_process(miss.name)
+            
+        # Checking that any valid observations are left after the get_obs_to_process function is run
+        if len(all_obs_info) == 0:
+            raise FileNotFoundError("No valid observations have been found, so cleanded_evt_lists may not be run.")
+
+        # all_obs_info is a list of lists, where each list is of the format: [ObsID, Inst, 'usable'].
+        #   There is a new list for each instrument, but I just want to loop over the ObsID in the following bit of code, 
+        #   I also want to know all the instruments that the ObsID contains events for
+        #   So here I am just making a dictionary of the format: {ObsID: insts}
+        # Getting unique obs_ids in all_obs_info
+        obs_ids = list(set([all_obs_info_list[0] for all_obs_info_list in all_obs_info]))
+        obs_info_dict = {}
+        for obs in obs_ids:
+            # Collecting all the insts that a certain ObsID has events for
+            insts = ''.join([all_obs_info_list[1] for all_obs_info_list in all_obs_info if obs in all_obs_info_list])
+            # The insts are all TM{x} where x is a number from 1-7, I just want to retain the x information, 
+            #   and append it to the dict
+            obs_info_dict[obs] = ''.join(ch for ch in insts if ch.isdigit())
+
+        # We iterate through the valid identifying information
+        for obs_id in obs_info_dict:
+
+            # Checking that flaregti has been run successfully on this observation so that it can be cleaned
+            if not obs_archive.process_success[miss.name]['flaregti'][obs_id]:
```

Comment:
You should use the archive "check_dependence_success" method for this, see the XMM processing functions for examples on usage (some of the different processing functions use it in slightly different ways, so I'd take a look at the usage in all of them). 

Again I tried to design it generally for different missions, but I don't know if it'll actually work, so just give it a go and give me a shout.

If you take a look at the actual check_dependence_success method you'll see what it does, but basically it's a centralised way to do what you're doing here.

Reply:
If you think this through, this function wouldn't work if even a single previous step had failed for _any_ of the eROSITA observations in the archive. The check_dependence_success method works in a slightly subtler way, just recording failures and avoiding those observations, and only erroring if **every** past step has failed.